src/sync/SSEHandler/NotificationKeeper.js

```diff
@@ -0,0 +1,56 @@
+import { PUSH_CONNECT, PUSH_DISCONNECT, PUSH_DISABLED, ControlTypes } from '../constants';
+
+const controlPriMatcher = /control_pri$/;
+
+export default function notificationKeeperFactory(feedbackLoopEmitter) {
+
+  let isStreamingUp;
+  let occupancyTimestamp = -1;
+  let controlTimestamp = -1;
+
+  return {
+    handleOpen() {
+      isStreamingUp = true;
+      feedbackLoopEmitter.emit(PUSH_CONNECT);
+    },
+
+    handleOccupancyEvent(publishers, channel, timestamp) {
+      if (controlPriMatcher.test(channel) && timestamp > occupancyTimestamp) {
+        occupancyTimestamp = timestamp;
+        if (publishers === 0 && isStreamingUp) {
+          isStreamingUp = false;
+          feedbackLoopEmitter.emit(PUSH_DISCONNECT); // notify(STREAMING_DOWN) in spec
+          return;
+        }
+        if (publishers !== 0 && !isStreamingUp) {
+          isStreamingUp = true;
+          feedbackLoopEmitter.emit(PUSH_CONNECT); // notify(STREAMING_UP) in spec
+        }
+      }
+    },
+
+    handleControlEvent(controlType, channel, timestamp) {
+      if (controlPriMatcher.test(channel) && timestamp > controlTimestamp) {
+        controlTimestamp = timestamp;
+        if (controlType === ControlTypes.STREAMING_PAUSED && isStreamingUp) {
+          isStreamingUp = false;
+          feedbackLoopEmitter.emit(PUSH_DISCONNECT); // notify(STREAMING_DOWN) in spec
+          return;
+        }
+        if (controlType === ControlTypes.STREAMING_RESUMED && !isStreamingUp) {
+          isStreamingUp = true;
+          feedbackLoopEmitter.emit(PUSH_CONNECT); // notify(STREAMING_UP) in spec
+          return;
+        }
+        if (controlType === ControlTypes.STREAMING_DISABLED) {
+          isStreamingUp = false;
+          feedbackLoopEmitter.emit(PUSH_DISABLED);
```

Comment:
Consider using else ifs to control the flow. Try not to think of returns as a GOTO, but a way to break execution. If all you need is for stopping after one condition was met, use a switch or if elseif

_If you love your returns, don't just leave it out because the current code works with that. If later someone else comes, adds a new type and misses the lack of return (and don't test properly of course), a bug might be introduced._